google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java

```diff
@@ -69,6 +109,266 @@ public String getDisplayName() {
+      this.id = instanceConfigInfo.id;
+      this.displayName = instanceConfigInfo.displayName;
+      this.replicas = instanceConfigInfo.replicas;
+      this.leaderOptions = instanceConfigInfo.leaderOptions;
+      this.optionalReplicas = instanceConfigInfo.optionalReplicas;
+      this.baseConfig = instanceConfigInfo.baseConfig;
+      this.configType = instanceConfigInfo.configType;
+      this.etag = instanceConfigInfo.etag;
+      this.reconciling = instanceConfigInfo.reconciling;
+      this.state = instanceConfigInfo.state;
+      this.labels = new HashMap<>(instanceConfigInfo.labels);
+    }
+
+    private void addAllReplicasFromBaseConfig() {
+      // Custom configurations contain all the replicas of the base config and at least one optional
+      // replica.
+      this.replicas = this.baseConfig.getReplicas();
+      this.replicas.add(baseConfig.getOptionalReplicas().get(0));
+    }
+
+    @Override
+    public Builder setDisplayName(String displayName) {
+      this.displayName = displayName;
+      return this;
+    }
+
+    @Override
+    public Builder addAllReplicas(List<ReplicaInfo> replicas) {
+      this.replicas = replicas;
+      return this;
+    }
+
+    @Override
+    public Builder addAllLeaderOptions(List<String> leaderOptions) {
+      this.leaderOptions = leaderOptions;
+      return this;
+    }
+
+    @Override
+    protected Builder addAllOptionalReplicas(List<ReplicaInfo> optionalReplicas) {
+      this.optionalReplicas = optionalReplicas;
+      return this;
+    }
+
+    @Override
+    public Builder setBaseConfig(InstanceConfigInfo baseConfig) {
+      this.baseConfig = baseConfig;
+      return this;
+    }
+
+    @Override
+    protected Builder setConfigType(Type configType) {
+      this.configType = configType;
+      return this;
+    }
+
+    @Override
+    protected Builder setState(State state) {
+      this.state = state;
+      return this;
+    }
+
+    @Override
+    public Builder setEtag(String etag) {
+      this.etag = etag;
+      return this;
+    }
+
+    @Override
+    protected Builder setReconciling(boolean reconciling) {
+      this.reconciling = reconciling;
+      return this;
+    }
+
+    @Override
+    public Builder addLabel(String key, String value) {
+      this.labels.put(key, value);
+      return this;
+    }
+
+    @Override
+    public Builder putAllLabels(Map<String, String> labels) {
+      this.labels.putAll(labels);
+      return this;
+    }
+
+    @Override
+    public Builder addReadOnlyReplicas(List<ReplicaInfo> readOnlyReplicas) {
+      this.replicas.addAll(readOnlyReplicas);
+      return this;
+    }
+
+    @Override
+    public InstanceConfigInfo build() {
+      return new InstanceConfigInfo(this);
+    }
+  }
+
+  public InstanceConfigInfo(InstanceConfigId id, String displayName) {
+    this((BuilderImpl) newBuilder(id).setDisplayName(displayName));
+  }
+
+  public static Builder newBuilder(InstanceConfigId id) {
```

Comment:
I would suggest removing this (or otherwise making it package-private if we need it for internal reasons). BaseConfig is always required for custom InstanceConfigs.

Reply:
Will make this newBuilder private.

Reply:
This newBuilder is used to create InstanceConfigInfo object to be passed during updateInstanceConfig. We do not require a baseConfig in that case.